### ISSUE TITLE
BUG: `copy.deepcopy()` or `gdf.copy(deep=True)` doesn't copy my dict in `.attrs`

### ISSUE THREAD

**Issue by LucaMarconato** [geopandas/geopandas#2920]
- [X] I have checked that this issue has not already been reported.

- [X] (almost) I have confirmed this bug exists on the latest version of geopandas: **tested on 0.13, not on 0.13.2**

- [ ] (optional) I have confirmed this bug exists on the main branch of geopandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
    import geopandas as gpd
    from shapely.geometry import Point
    import copy

    # Create a GeoDataFrame
    gdf = gpd.GeoDataFrame({
        'geometry': [Point(1, 1), Point(2, 2)],
        'value': [1, 2]
    })

    # Add an attribute to the GeoDataFrame
    gdf.attrs['transform'] = {}

    # Copy the GeoDataFrame
    gdf2 = copy.deepcopy(gdf)
    # gdf2 = gdf.copy(deep=True)

    # Print the IDs
    print(id(gdf.attrs['transform']))  # ID of copied attribute
    print(id(gdf2.attrs['transform']))  # ID of original attribute

    # Assertion
    assert gdf.attrs['transform'] is not gdf2.attrs['transform']

```

#### Problem description

I would expect the dict in `.attrs` to be a copy since I performed a deep copy, but it is still the same object.

#### Expected Output

The assertion should not fail.

#### Output of ``geopandas.show_versions()``

```
SYSTEM INFO
-----------
python     : 3.10.0 (default, Mar  3 2022, 03:54:28) [Clang 12.0.0 ]
executable : /Users/macbook/miniconda3/envs/ome/bin/python
machine    : macOS-13.2.1-arm64-arm-64bit

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.11.1
GEOS lib   : None
GDAL       : 3.6.4
GDAL data dir: /Users/macbook/miniconda3/envs/ome/lib/python3.10/site-packages/fiona/gdal_data
PROJ       : 9.2.0
PROJ data dir: /Users/macbook/miniconda3/envs/ome/lib/python3.10/site-packages/pyproj/proj_dir/share/proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.13.0
numpy      : 1.24.3
pandas     : 2.0.2
pyproj     : 3.5.0
shapely    : 2.0.1
fiona      : 1.9.4.post1
geoalchemy2: None
geopy      : None
matplotlib : 3.7.1
mapclassify: None
pygeos     : 0.14
pyogrio    : None
psycopg2   : None
pyarrow    : 12.0.0
rtree      : None
```

</details>

**Comment by m-richards** [comment]
Hi @LucaMarconato, this appears to be an upstream issue in pandas, if I modify your example I can remove geopandas completely and still see the same behaviour:

```python
import pandas as pd
import copy

# Create a GeoDataFrame
df = pd.DataFrame({
    'value': [1, 2]
})

# Add an attribute to the DataFrame
df.attrs['transform'] = {}

# Copy the DataFrame
df2 = copy.deepcopy(df)
# gdf2 = gdf.copy(deep=True)

# Print the IDs
print(id(df.attrs['transform']))  # ID of copied attribute
print(id(df2.attrs['transform']))  # ID of original attribute

# Assertion
assert df.attrs['transform'] is not df2.attrs['transform']
```
I would suggest then it makes sense to report this issue on the pandas issue tracker as well and look for a fix there.

**Comment by m-richards** [comment]
Thanks @LucaMarconato for following through on this, I can see this is now been fixed in pandas 2.2 so will close.
